rust/hyperlane-base/src/contract_sync/mod.rs

```diff
@@ -1,31 +1,143 @@
+use std::{marker::PhantomData, sync::Arc};
+
 use derive_new::new;
 
-pub use cursor::*;
-use hyperlane_core::HyperlaneDomain;
-pub use interchain_gas::*;
-pub use mailbox::*;
+use cursor::*;
+use hyperlane_core::{
+    ContractSyncCursor, HyperlaneDB, HyperlaneDomain, HyperlaneHighWatermarkDB, HyperlaneMessage,
+    HyperlaneMessageDB, Indexer, MessageIndexer,
+};
 pub use metrics::ContractSyncMetrics;
+use std::fmt::Debug;
+use tracing::{debug, info};
 
-use crate::{chains::IndexSettings, db::HyperlaneDB};
+use crate::chains::IndexSettings;
 
 mod cursor;
 mod eta_calculator;
-mod interchain_gas;
-/// Tools for working with message continuity.
-pub mod last_message;
-mod mailbox;
 mod metrics;
-mod schema;
 
+// Okay, how do I allow
 /// Entity that drives the syncing of an agent's db with on-chain data.
 /// Extracts chain-specific data (emitted checkpoints, messages, etc) from an
 /// `indexer` and fills the agent's db with this data. A CachingMailbox
 /// will use a contract sync to spawn syncing tasks to keep the db up-to-date.
-#[derive(Debug, new)]
-pub(crate) struct ContractSync<I> {
+#[derive(Debug, new, Clone)]
+pub struct ContractSync<T, D: HyperlaneDB<T>, I: Indexer<T>> {
     domain: HyperlaneDomain,
-    db: HyperlaneDB,
+    db: D,
     indexer: I,
-    index_settings: IndexSettings,
     metrics: ContractSyncMetrics,
+    // TODO: Why is this necessary?
+    blah: PhantomData<T>,
```

Comment:
@mattiecnvr help please, the compiler is otherwise unhappy with
```
pub struct ContractSync<T, D: HyperlaneDB<T>, I: Indexer<T>>
```

I'm using `D` and `I` so that the impl can be aware of exactly what type of indexer or DB it has, which can be useful for creating special cursors.

There's probably a better way that doesn't involve the cursors being made by the ContractSync, but I'd refactored enough at this point that I just wanted to roll with what I had.

Reply:
You could move the T type into the function generics instead. The reason you would need `PhatomData` generically speaking is if you have a type that is not actually represented in the trait/struct.

Reply:
How would that work? I assume I needed it in the struct so as to restrict the D/I types.